File: bindings/kotlin/src/main/kotlin/dev/qperiapt/QPeriaptHybrid.kt
package dev.qperiapt

import java.lang.foreign.Arena
import java.lang.foreign.FunctionDescriptor
import java.lang.foreign.Linker
import java.lang.foreign.MemorySegment
import java.lang.foreign.SymbolLookup
import java.lang.foreign.ValueLayout.ADDRESS
import java.lang.foreign.ValueLayout.JAVA_BYTE
import java.lang.foreign.ValueLayout.JAVA_INT
import java.lang.foreign.ValueLayout.JAVA_LONG
import java.nio.file.Files
import java.nio.file.Path

/**
 * Kotlin product face of the PQ/T hybrid suite over the ABI-major C library, via the Foreign
 * Function & Memory API (Project Panama, JDK 22+). It reuses the Rust core but obtains
 * randomness inside the native ABI, so tests assert semantic parity rather than deterministic
 * byte replay — see bindings/README.md.
 *
 * Returned secret arrays are caller-owned. Invoke [EncapsulationResult.wipeSecret]
 * or `fill(0)` on every secret/key array after use; JVM and OS copies remain outside
 * this binding's control.
 */
object QPeriaptHybrid {
    const val ABI_VERSION = 2
    const val PROFILE_CONTEXT_BOUND: Byte = 2
    const val SECRET_LEN = 32
    const val MLKEM_PK_LEN = 1184
    const val MLKEM_SK_LEN = 2400
    const val MLKEM_CT_LEN = 1088
    const val X25519_LEN = 32
    const val POLICY_DECISION_LEN = 40
    const val TRUSTED_POLICY_STATE_LEN = 36
    const val MAX_SIGNED_POLICY_BYTES = 64 * 1024
    const val MAX_APPLICATION_CONTEXT_BYTES = 64 * 1024
    const val SUITE_MLKEM768_X25519: Byte = 1
    const val KEY_FORMAT_EXPANDED: Byte = 1

    class QPeriaptException(val operation: String, val code: Int) :
        RuntimeException("$operation rc=$code")

    data class EncapsulationResult(
        val ctPq: ByteArray,
        val ctTrad: ByteArray,
        /** Caller-owned secret material. Wipe this array when the session no longer needs it. */
        val secret: ByteArray,
    ) {
        /** Best-effort zeroization of this result's current secret array. */
        fun wipeSecret() {
            secret.fill(0)
        }
    }

    data class KeyPairResult(
        val skPq: ByteArray,
        val pkPq: ByteArray,
        val skTrad: ByteArray,
        val pkTrad: ByteArray,
    ) {
        fun wipeSecrets() {
            skPq.fill(0)
            skTrad.fill(0)
        }
    }

    /** Atomic suite/profile/version decision produced by a verified signed policy. */
    class PolicyDecision internal constructor(encoded: ByteArray) {
        val suiteCode: Byte
        val profile: Byte
        val keyFormat: Byte
        val policyVersion: Long
        private val digest: ByteArray
        private val canonical: ByteArray = encoded.clone()

        init {
            require(encoded.size == POLICY_DECISION_LEN) { "invalid policy decision length" }
            require(encoded[0] == 1.toByte()) { "unknown policy decision version" }
            require(encoded[1] == SUITE_MLKEM768_X25519) { "unsupported policy suite" }
            require(encoded[2] == PROFILE_CONTEXT_BOUND) { "unsupported policy profile" }
            require(encoded[3] == KEY_FORMAT_EXPANDED) { "unsupported policy key format" }
            suiteCode = encoded[1]
            profile = encoded[2]
            keyFormat = encoded[3]
            policyVersion = encoded.copyOfRange(4, 8).fold(0L) { value, byte ->
                (value shl 8) or (byte.toLong() and 0xff)
            }
            require(policyVersion != 0L) { "zero policy version" }
            digest = encoded.copyOfRange(8, POLICY_DECISION_LEN)
        }

        fun policyDigest(): ByteArray = digest.clone()

        internal fun encoded(): ByteArray = canonical.clone()

        /** Persist atomically and pass to the next decision load. */
        fun trustedState(): ByteArray = byteArrayOf(
            (policyVersion ushr 24).toByte(),
            (policyVersion ushr 16).toByte(),
            (policyVersion ushr 8).toByte(),
            policyVersion.toByte(),
        ) + digest
    }

    private val linker = Linker.nativeLinker()
    // Product code must bind a specific native library path, then verify ABI/suite metadata below.
    private val lookup: SymbolLookup = run {
        val explicit = System.getProperty("qperiapt.lib")
            ?: error("qperiapt.lib must be set to an absolute q-periapt native library path")
        val path = Path.of(explicit)
        require(path.isAbsolute) { "qperiapt.lib must be an absolute path: $explicit" }
        require(Files.isRegularFile(path)) { "qperiapt.lib does not name a regular file: $explicit" }
        SymbolLookup.libraryLookup(path, Arena.global())
    }

    private fun handle(name: String, desc: FunctionDescriptor) =
        linker.downcallHandle(lookup.find(name).orElseThrow { RuntimeException("missing symbol $name") }, desc)

    private val abiVersionFn = handle("q_periapt_abi_version", FunctionDescriptor.of(JAVA_INT))
    private val versionFn = handle("q_periapt_version", FunctionDescriptor.of(ADDRESS))
    private val fixedSuiteIdFn = handle("q_periapt_fixed_suite_id", FunctionDescriptor.of(ADDRESS))
    private val fixedSuiteIdLenFn = handle("q_periapt_fixed_suite_id_len", FunctionDescriptor.of(JAVA_LONG))
    private val statusNameFn = handle("q_periapt_status_name", FunctionDescriptor.of(ADDRESS, JAVA_INT))

    private val decisionPolicyDesc = FunctionDescriptor.of(
        JAVA_INT,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG
    )
    private val decisionPolicy =
        handle("q_periapt_decision_from_signed_policy", decisionPolicyDesc)

    // (ptr,long) buffer pairs are modeled as ADDRESS + JAVA_LONG.
    private val generateKeypairDesc = FunctionDescriptor.of(
        JAVA_INT,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG
    )
    private val generateKeypair = handle("q_periapt_generate_keypair", generateKeypairDesc)

    private val encapDesc = FunctionDescriptor.of(
        JAVA_INT,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG,
        ADDRESS, JAVA_LONG
    )
    private val encap = handle("q_periapt_encapsulate", encapDesc)

    private val decapDesc = FunctionDescriptor.of(
        JAVA_INT,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG,
        ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG, ADDRESS, JAVA_LONG
    )
    private val decap = handle("q_periapt_decapsulate", decapDesc)

    private fun Arena.seg(bytes: ByteArray): MemorySegment =
        allocate(bytes.size.toLong().coerceAtLeast(1)).also {
            MemorySegment.copy(bytes, 0, it, JAVA_BYTE, 0, bytes.size)
        }

    /**
     * Owns native segments that contain secret material from the moment they are
     * allocated. The owner wipes every registered segment before the surrounding
     * arena releases it, including allocation/copy/native-call failure paths.
     */
    private class SecretSegments(private val arena: Arena) : AutoCloseable {
        private val segments = ArrayList<MemorySegment>()

        fun copy(bytes: ByteArray): MemorySegment =
            allocate(bytes.size.toLong().coerceAtLeast(1)).also {
                MemorySegment.copy(bytes, 0, it, JAVA_BYTE, 0, bytes.size)
            }

        fun allocate(byteSize: Long): MemorySegment =
            arena.allocate(byteSize.coerceAtLeast(1)).also { segments.add(it) }

        override fun close() {
            var failure: Throwable? = null
            for (segment in segments.asReversed()) {
                try {
                    segment.fill(0)
                } catch (error: Throwable) {
                    val first = failure
                    if (first == null) {
                        failure = error
                    } else {
                        first.addSuppressed(error)
                    }
                }
            }
            failure?.let { throw it }
        }
    }

    private fun MemorySegment.toBytes(n: Int): ByteArray =
        ByteArray(n).also { MemorySegment.copy(this, JAVA_BYTE, 0, it, 0, n) }

    // A confined Arena frees its native memory on close but does NOT zero it; explicitly
    // wipe any segment that held secret material (keys, shared secrets, combiner inputs)
    // before that happens, so freed pages cannot leak a secret to a later allocation.
    private fun MemorySegment.wipe() {
        fill(0)
    }

    private fun checkOk(operation: String, rc: Int) {
        if (rc != 0) throw QPeriaptException(operation, rc)
    }

    private fun cString(pointer: MemorySegment, maxBytes: Long): String {
        require(pointer != MemorySegment.NULL) { "native string pointer was NULL" }
        return pointer.reinterpret(maxBytes).getString(0)
    }

    fun runtimeAbiVersion(): Int = abiVersionFn.invokeExact() as Int

    fun runtimeVersion(): String = cString(versionFn.invokeExact() as MemorySegment, 64)

    fun fixedSuiteId(): ByteArray = cString(fixedSuiteIdFn.invokeExact() as MemorySegment, 64).encodeToByteArray()

    fun fixedSuiteIdLen(): Long = fixedSuiteIdLenFn.invokeExact() as Long

    fun statusName(code: Int): String = cString(statusNameFn.invokeExact(code) as MemorySegment, 64)

    private fun validateRuntimeMetadata() {
        check(runtimeAbiVersion() == ABI_VERSION) {
            "q-periapt ABI mismatch: header=$ABI_VERSION runtime=${runtimeAbiVersion()}"
        }
        check(fixedSuiteId().contentEquals("ML-KEM-768+X25519".encodeToByteArray())) {
            "q-periapt fixed suite mismatch: ${fixedSuiteId().decodeToString()}"
        }
        check(fixedSuiteIdLen() == "ML-KEM-768+X25519".length.toLong()) {
            "q-periapt fixed suite length mismatch: ${fixedSuiteIdLen()}"
        }
    }

    init {
        validateRuntimeMetadata()
    }

    fun decisionFromSignedPolicy(
        toml: ByteArray,
        signature: ByteArray,
        verificationKey: ByteArray,
        lastTrustedState: ByteArray = byteArrayOf(),
    ): PolicyDecision = Arena.ofConfined().use { a ->
        require(toml.size <= MAX_SIGNED_POLICY_BYTES) {
            "toml exceeds $MAX_SIGNED_POLICY_BYTES bytes: ${toml.size}"
        }
        require(lastTrustedState.isEmpty() || lastTrustedState.size == TRUSTED_POLICY_STATE_LEN) {
            "lastTrustedState must be empty or $TRUSTED_POLICY_STATE_LEN bytes"
        }
        val tomlSeg = a.seg(toml)
        val sigSeg = a.seg(signature)
        val vkSeg = a.seg(verificationKey)
        val stateSeg = a.seg(lastTrustedState)
        val out = a.allocate(POLICY_DECISION_LEN.toLong())
        try {
            val rc = decisionPolicy.invokeExact(
                tomlSeg, toml.size.toLong(), sigSeg, signature.size.toLong(),
                vkSeg, verificationKey.size.toLong(),
                stateSeg, lastTrustedState.size.toLong(),
                out, POLICY_DECISION_LEN.toLong()
            ) as Int
            checkOk("q_periapt_decision_from_signed_policy", rc)
            PolicyDecision(out.toBytes(POLICY_DECISION_LEN))
        } finally {
            stateSeg.wipe(); out.wipe()
        }
    }

    fun generateKeypair(decision: PolicyDecision): KeyPairResult = Arena.ofConfined().use { a ->
        SecretSegments(a).use { secrets ->
            val skPq = secrets.allocate(MLKEM_SK_LEN.toLong())
            val pkPq = a.allocate(MLKEM_PK_LEN.toLong())
            val skTrad = secrets.allocate(X25519_LEN.toLong())
            val pkTrad = a.allocate(X25519_LEN.toLong())
            val decisionBytes = decision.encoded()
            val rc = generateKeypair.invokeExact(
                a.seg(decisionBytes), decisionBytes.size.toLong(),
                skPq, MLKEM_SK_LEN.toLong(), pkPq, MLKEM_PK_LEN.toLong(),
                skTrad, X25519_LEN.toLong(), pkTrad, X25519_LEN.toLong()
            ) as Int
            checkOk("q_periapt_generate_keypair", rc)
            KeyPairResult(
                skPq = skPq.toBytes(MLKEM_SK_LEN),
                pkPq = pkPq.toBytes(MLKEM_PK_LEN),
                skTrad = skTrad.toBytes(X25519_LEN),
                pkTrad = pkTrad.toBytes(X25519_LEN),
            )
        }
    }

    /** Encapsulate while committing the exact signed-policy digest and application context. */
    fun encapsulate(
        decision: PolicyDecision,
        pkPq: ByteArray,
        pkTrad: ByteArray,
        applicationContext: ByteArray,
    ): EncapsulationResult = Arena.ofConfined().use { a ->
        require(applicationContext.size <= MAX_APPLICATION_CONTEXT_BYTES) {
            "applicationContext exceeds $MAX_APPLICATION_CONTEXT_BYTES bytes: ${applicationContext.size}"
        }
        SecretSegments(a).use { secrets ->
            val decisionBytes = decision.encoded()
            val decisionSeg = a.seg(decisionBytes)
            val outCtPq = a.allocate(MLKEM_CT_LEN.toLong())
            val outCtTrad = a.allocate(X25519_LEN.toLong())
            val outSecret = secrets.allocate(SECRET_LEN.toLong())
            val rc = encap.invokeExact(
                decisionSeg, decisionBytes.size.toLong(),
                a.seg(pkPq), pkPq.size.toLong(), a.seg(pkTrad), pkTrad.size.toLong(),
                a.seg(applicationContext), applicationContext.size.toLong(),
                outCtPq, MLKEM_CT_LEN.toLong(), outCtTrad, X25519_LEN.toLong(),
                outSecret, SECRET_LEN.toLong()
            ) as Int
            checkOk("q_periapt_encapsulate", rc)
            EncapsulationResult(
                ctPq = outCtPq.toBytes(MLKEM_CT_LEN),
                ctTrad = outCtTrad.toBytes(X25519_LEN),
                secret = outSecret.toBytes(SECRET_LEN),
            )
        }
    }

    /** Decapsulate under the same policy decision and application context. */
    fun decapsulate(
        decision: PolicyDecision,
        skPq: ByteArray,
        ctPq: ByteArray,
        pkPq: ByteArray,
        skTrad: ByteArray,
        ctTrad: ByteArray,
        pkTrad: ByteArray,
        applicationContext: ByteArray,
    ): ByteArray = Arena.ofConfined().use { a ->
        require(applicationContext.size <= MAX_APPLICATION_CONTEXT_BYTES) {
            "applicationContext exceeds $MAX_APPLICATION_CONTEXT_BYTES bytes: ${applicationContext.size}"
        }
        SecretSegments(a).use { secrets ->
            val decisionBytes = decision.encoded()
            val decisionSeg = a.seg(decisionBytes)
            val skPqSeg = secrets.copy(skPq)
            val skTradSeg = secrets.copy(skTrad)
            val out = secrets.allocate(SECRET_LEN.toLong())
            val rc = decap.invokeExact(
                decisionSeg, decisionBytes.size.toLong(),
                skPqSeg, skPq.size.toLong(), a.seg(ctPq), ctPq.size.toLong(),
                a.seg(pkPq), pkPq.size.toLong(), skTradSeg, skTrad.size.toLong(),
                a.seg(ctTrad), ctTrad.size.toLong(), a.seg(pkTrad), pkTrad.size.toLong(),
                a.seg(applicationContext), applicationContext.size.toLong(),
                out, SECRET_LEN.toLong()
            ) as Int
            checkOk("q_periapt_decapsulate", rc)
            out.toBytes(SECRET_LEN)
        }
    }

}

File: bindings/swift/Sources/QPeriaptHybrid/QPeriaptHybrid.swift
// SwiftPM imports the C ABI through the CQPeriapt system-library target. The iOS/macOS device
// runner imports the same generated C header through a bridging header, where this module is not
// present but the declarations are already in scope.
#if canImport(CQPeriapt)
import CQPeriapt
#endif
import Foundation

/// Combiner profile selector (mirrors `Q_PERIAPT_PROFILE_*`).
public enum QPeriaptProfile: UInt8 {
    case contextBound = 2
}

/// Atomic decision produced by a verified signed policy for the fixed native suite.
/// Fields are read-only so Swift callers cannot accidentally mix a profile from one
/// policy with a suite/version from another.
public struct QPeriaptPolicyDecision {
    public let suiteCode: UInt8
    public let profile: QPeriaptProfile
    public let keyFormatCode: UInt8
    public let policyVersion: UInt32
    public let policyDigest: [UInt8]
    fileprivate let encoded: [UInt8]

    /// Persist this exact value atomically after accepting the policy and pass it
    /// to the next ``QPeriaptHybrid/decisionFromSignedPolicy`` call.
    public var trustedState: [UInt8] {
        var state = Array(policyVersion.bigEndianBytes)
        state.append(contentsOf: policyDigest)
        return state
    }

    fileprivate init(encoded: [UInt8]) throws {
        guard encoded.count == Int(Q_PERIAPT_POLICY_DECISION_LEN),
              encoded[0] == UInt8(Q_PERIAPT_POLICY_DECISION_VERSION),
              encoded[1] == UInt8(Q_PERIAPT_SUITE_MLKEM768_X25519),
              let profile = QPeriaptProfile(rawValue: encoded[2]),
              encoded[3] == UInt8(Q_PERIAPT_KEY_FORMAT_EXPANDED)
        else {
            throw QPeriaptError(code: Q_PERIAPT_ERR_POLICY)
        }
        let version = encoded[4..<8].reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
        guard version != 0 else { throw QPeriaptError(code: Q_PERIAPT_ERR_POLICY) }
        let keyFormat = encoded[3]
        guard profile == .contextBound,
              keyFormat == UInt8(Q_PERIAPT_KEY_FORMAT_EXPANDED) else {
            throw QPeriaptError(code: Q_PERIAPT_ERR_POLICY)
        }
        self.suiteCode = encoded[1]
        self.profile = profile
        self.keyFormatCode = keyFormat
        self.policyVersion = version
        self.policyDigest = Array(encoded[8...])
        self.encoded = encoded
    }
}

private extension UInt32 {
    var bigEndianBytes: [UInt8] {
        let value = bigEndian
        return withUnsafeBytes(of: value) { Array($0) }
    }
}

/// Errors mirroring the C status codes.
public struct QPeriaptError: Error {
    public static let lengthCode = Int32(Q_PERIAPT_ERR_LENGTH)
    public static let policyCode = Int32(Q_PERIAPT_ERR_POLICY)

    public let code: Int32
}

/// Result of hybrid encapsulation.
public struct QPeriaptEncapsulation {
    public let ctPq: [UInt8]
    public let ctTrad: [UInt8]
    public private(set) var secret: [UInt8]

    /// Best-effort destruction of this value's current secret buffer.
    ///
    /// Swift arrays use copy-on-write. Callers must also wipe every independent
    /// copy they created; this method cannot erase copies held elsewhere or by
    /// the operating system.
    public mutating func wipeSecret() {
        QPeriaptHybrid.wipe(&secret)
    }
}

/// One atomically generated fixed-suite key pair. Secret-key buffers are
/// caller-owned and must be wiped when no longer needed.
public struct QPeriaptKeyPair {
    public private(set) var skPq: [UInt8]
    public let pkPq: [UInt8]
    public private(set) var skTrad: [UInt8]
    public let pkTrad: [UInt8]

    public mutating func wipeSecrets() {
        QPeriaptHybrid.wipe(&skPq)
        QPeriaptHybrid.wipe(&skTrad)
    }
}

/// Swift face of the PQ/T hybrid suite (ML-KEM-768 + X25519), over the C ABI.
///
/// Secret hygiene: returned secrets (`sk`, the decapsulated/combined secret) are caller-owned
/// Swift `[UInt8]` value arrays. Because Swift arrays are copy-on-write, the binding cannot zero
/// them after `return` without corrupting the value it hands back, so wiping is the caller's
/// responsibility — call ``wipe(_:)`` on each returned array and ``QPeriaptEncapsulation/wipeSecret()``
/// on every encapsulation value when finished with it. These APIs are best-effort: they cannot erase
/// independent copy-on-write buffers or operating-system copies.
public enum QPeriaptHybrid {
    public static let abiVersion = UInt32(Q_PERIAPT_ABI_VERSION)
    public static let secretLen = Int(Q_PERIAPT_SECRET_LEN)
    public static let mlkemPkLen = Int(Q_PERIAPT_MLKEM768_PK_LEN)
    public static let mlkemSkLen = Int(Q_PERIAPT_MLKEM768_SK_LEN)
    public static let mlkemCtLen = Int(Q_PERIAPT_MLKEM768_CT_LEN)
    public static let x25519Len = Int(Q_PERIAPT_X25519_LEN)
    public static let policyDecisionLen = Int(Q_PERIAPT_POLICY_DECISION_LEN)
    public static let trustedPolicyStateLen = Int(Q_PERIAPT_TRUSTED_POLICY_STATE_LEN)
    public static let maxSignedPolicyBytes = Int(Q_PERIAPT_MAX_SIGNED_POLICY_BYTES)
    public static let maxApplicationContextBytes = Int(Q_PERIAPT_MAX_APPLICATION_CONTEXT_BYTES)
    public static let fixedSuiteCode = UInt8(Q_PERIAPT_SUITE_MLKEM768_X25519)

    public static var runtimeAbiVersion: UInt32 {
        q_periapt_abi_version()
    }

    public static var runtimeVersion: String {
        String(cString: q_periapt_version())
    }

    public static var fixedSuiteId: [UInt8] {
        Array(String(cString: q_periapt_fixed_suite_id()).utf8)
    }

    public static var fixedSuiteIdLen: Int {
        Int(q_periapt_fixed_suite_id_len())
    }

    public static func statusName(_ code: Int32) -> String {
        String(cString: q_periapt_status_name(code))
    }

    public static func decisionFromSignedPolicy(
        toml: [UInt8], signature: [UInt8], verificationKey: [UInt8],
        lastTrustedState: [UInt8] = []
    ) throws -> QPeriaptPolicyDecision {
        guard lastTrustedState.isEmpty
                || lastTrustedState.count == Int(Q_PERIAPT_TRUSTED_POLICY_STATE_LEN)
        else {
            throw QPeriaptError(code: Q_PERIAPT_ERR_LENGTH)
        }
        var encoded = [UInt8](
            repeating: 0, count: Int(Q_PERIAPT_POLICY_DECISION_LEN))
        let rc = q_periapt_decision_from_signed_policy(
            toml, UInt(toml.count), signature, UInt(signature.count),
            verificationKey, UInt(verificationKey.count),
            lastTrustedState, UInt(lastTrustedState.count),
            &encoded, UInt(encoded.count))
        guard rc == Q_PERIAPT_OK else { throw QPeriaptError(code: rc) }
        return try QPeriaptPolicyDecision(encoded: encoded)
    }

    public static func generateKeypair(
        decision: QPeriaptPolicyDecision
    ) throws -> QPeriaptKeyPair {
        var skPq = [UInt8](repeating: 0, count: mlkemSkLen)
        var pkPq = [UInt8](repeating: 0, count: mlkemPkLen)
        var skTrad = [UInt8](repeating: 0, count: x25519Len)
        var pkTrad = [UInt8](repeating: 0, count: x25519Len)
        let rc = q_periapt_generate_keypair(
            decision.encoded, UInt(decision.encoded.count),
            &skPq, UInt(skPq.count), &pkPq, UInt(pkPq.count),
            &skTrad, UInt(skTrad.count), &pkTrad, UInt(pkTrad.count))
        guard rc == Q_PERIAPT_OK else { throw QPeriaptError(code: rc) }
        return QPeriaptKeyPair(skPq: skPq, pkPq: pkPq, skTrad: skTrad, pkTrad: pkTrad)
    }

    /// Encapsulate under one authenticated policy decision. The native core
    /// canonically commits the exact policy digest and `applicationContext`;
    /// a CompatXWing decision is rejected because that profile ignores context.
    public static func encapsulate(
        decision: QPeriaptPolicyDecision,
        pkPq: [UInt8], pkTrad: [UInt8], applicationContext: [UInt8]
    ) throws -> QPeriaptEncapsulation {
        var ctPq = [UInt8](repeating: 0, count: mlkemCtLen)
        var ctTrad = [UInt8](repeating: 0, count: x25519Len)
        var secret = [UInt8](repeating: 0, count: secretLen)
        let rc = q_periapt_encapsulate(
            decision.encoded, UInt(decision.encoded.count),
            pkPq, UInt(pkPq.count), pkTrad, UInt(pkTrad.count),
            applicationContext, UInt(applicationContext.count),
            &ctPq, UInt(ctPq.count), &ctTrad, UInt(ctTrad.count),
            &secret, UInt(secret.count))
        guard rc == Q_PERIAPT_OK else { throw QPeriaptError(code: rc) }
        return QPeriaptEncapsulation(ctPq: ctPq, ctTrad: ctTrad, secret: secret)
    }

    /// Decapsulate under the same authenticated decision and application context
    /// used by the policy-controlled encapsulation path.
    public static func decapsulate(
        decision: QPeriaptPolicyDecision,
        skPq: [UInt8], ctPq: [UInt8], pkPq: [UInt8],
        skTrad: [UInt8], ctTrad: [UInt8], pkTrad: [UInt8],
        applicationContext: [UInt8]
    ) throws -> [UInt8] {
        var secret = [UInt8](repeating: 0, count: secretLen)
        let rc = q_periapt_decapsulate(
            decision.encoded, UInt(decision.encoded.count),
            skPq, UInt(skPq.count), ctPq, UInt(ctPq.count), pkPq, UInt(pkPq.count),
            skTrad, UInt(skTrad.count), ctTrad, UInt(ctTrad.count),
            pkTrad, UInt(pkTrad.count),
            applicationContext, UInt(applicationContext.count),
            &secret, UInt(secret.count))
        guard rc == Q_PERIAPT_OK else { throw QPeriaptError(code: rc) }
        return secret
    }

    /// Securely zero a buffer that held secret material. Uses `memset_s`, which (unlike a plain
    /// `memset` of a soon-to-be-freed buffer) the compiler is not permitted to elide as a dead
    /// store. Call this on any `sk` or session secret once it is no longer needed.
    public static func wipe(_ buffer: inout [UInt8]) {
        buffer.withUnsafeMutableBytes { raw in
            guard let base = raw.baseAddress, raw.count > 0 else { return }
            memset_s(base, raw.count, 0, raw.count)
        }
    }
}
